core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -348,54 +348,59 @@ String doFateOperation(FateOperation op, List<ByteBuffer> args, Map<String,Strin
-              "Target namespace does not exist");
-        default:
-          String tableInfo = Tables.getPrintableTableInfoFromName(context, tableOrNamespaceName);
-          throw new AccumuloSecurityException(e.user, e.code, tableInfo, e);
-      }
-    } catch (ThriftTableOperationException e) {
-      switch (e.getType()) {
-        case EXISTS:
-          throw new TableExistsException(e);
-        case NOTFOUND:
-          throw new TableNotFoundException(e);
-        case NAMESPACE_EXISTS:
-          throw new NamespaceExistsException(e);
-        case NAMESPACE_NOTFOUND:
-          throw new NamespaceNotFoundException(e);
-        case OFFLINE:
-          throw new TableOfflineException(
-              Tables.getTableOfflineMsg(context, Tables.getTableId(context, tableOrNamespaceName)));
-        default:
-          throw new AccumuloException(e.description, e);
-      }
-    } catch (Exception e) {
-      throw new AccumuloException(e.getMessage(), e);
-    } finally {
-      Tables.clearCache(context);
-      // always finish table op, even when exception
-      if (opid != null)
-        try {
-          finishFateOperation(opid);
-        } catch (Exception e) {
-          log.warn("Exception thrown while finishing fate table operation", e);
+    // keep retrying bulk import if a concurrent merge happens. all other ops throw error or return
+    while (true) {
+      try {
+        opid = beginFateOperation();
+        executeFateOperation(opid, op, args, opts, !wait);
+        if (!wait) {
+          opid = null;
+          return null;
+        }
+        return waitForFateOperation(opid);
+      } catch (ThriftSecurityException e) {
+        switch (e.getCode()) {
+          case TABLE_DOESNT_EXIST:
+            throw new TableNotFoundException(null, tableOrNamespaceName,
+                "Target table does not exist");
+          case NAMESPACE_DOESNT_EXIST:
+            throw new NamespaceNotFoundException(null, tableOrNamespaceName,
+                "Target namespace does not exist");
+          default:
+            String tableInfo = Tables.getPrintableTableInfoFromName(context, tableOrNamespaceName);
+            throw new AccumuloSecurityException(e.user, e.code, tableInfo, e);
         }
+      } catch (ThriftTableOperationException e) {
+        switch (e.getType()) {
+          case EXISTS:
+            throw new TableExistsException(e);
+          case NOTFOUND:
+            throw new TableNotFoundException(e);
+          case NAMESPACE_EXISTS:
+            throw new NamespaceExistsException(e);
+          case NAMESPACE_NOTFOUND:
+            throw new NamespaceNotFoundException(e);
+          case OFFLINE:
+            throw new TableOfflineException(Tables.getTableOfflineMsg(context,
+                Tables.getTableId(context, tableOrNamespaceName)));
+          case BULK_CONCURRENT_MERGE:
+            log.info("Concurrent merge happened. Retrying Bulk Import to " + e.tableName);
+            break;
```

Comment:
I think this is the wrong place to retry.  I think the load plan needs to be recomputed.  Could throw an internal (non-API) runtime exception here and catch that exception in the bulk code.

Reply:
Ah yeah... good point about the plan being recomputed.  How do we handle the case where the user generates the plan?  I guess the plan will be different if we generate it, but the user plan will still be the same.

Reply:
> How do we handle the case where the user generates the plan?

Depends on the range types in the LoadPlan the user provided.  If the RangeType is TABLET, then an error needs to be thrown.  If the RangeType is FILE, then those could be recomputed.

Reply:
OK thanks.  I think this is a good spot to throw a custom exception though, based on this Thrift type.